source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -431,15 +431,41 @@ Network::FilterStatus Filter::initializeUpstreamConnection() {
       return Network::FilterStatus::StopIteration;
     }
   } else {
-    Http::ConnectionPool::Instance* conn_pool = cluster_manager_.httpConnPoolForCluster(
-        cluster_name, Upstream::ResourcePriority::Default, Http::Protocol::Http2, this);
-    if (conn_pool) {
-      upstream_ = std::make_unique<HttpUpstream>(*upstream_callbacks_,
-                                                 config_->tunnelingConfig()->hostname());
-      HttpUpstream* http_upstream = static_cast<HttpUpstream*>(upstream_.get());
-      upstream_handle_ = std::make_shared<HttpConnectionHandle>(
-          conn_pool->newStream(http_upstream->responseDecoder(), *this));
-      return Network::FilterStatus::StopIteration;
+    auto generic_pool = cluster_manager_.genericConnPoolForCluster(
```

Comment:
If we agree that endpoint metadata is the way to go (I think I'm OK with this as we have done this type of thing for TLS migrations in the past), my strong preference is to put all of that functionality here within TCP proxy vs. having so many cluster manager changes. IIRC you can get access to the cluster load balancer and have the LB pick a host. At that point, you can look at host metadata and then decide on HTTP or TCP for the given host?  It might require some small changes to allocate/fetch a conn pool for a specific host but that seems OK to me? Do we think that is possible?

This logic seems very specific in practice to TCP proxy so it seems odd that we would use this anywhere else.

Reply:
Thanks for the comment on metadata.
To me it's a tcpproxy internal thing at this moment. I am not ready to push to other terminate network filter. I will update the PR to reflect your opinion

Reply:
I'm good either way, but I would prefer that we not allow the mixed mode unless explicitly configured.  That would either be that we explicitly tag endpoints "ok to do raw TCP for connect" or we tag with capabilities (http/2, raw tcp) but only look at the metadata if the proxy_config says to select based on metadata.  Make sense?

Reply:
I will add the config in tcp proxy, something like "select_tunneling_using_endpoint_metadata" beside tunneling_config.
Or is the selector part of tunneling_config?